### PR TITLE
[USMON-910] Introduce postgres TLS monitoring

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Introducing the ability to capture and encrypted postgres traffic (a.k.a postgres TLS).
The change uses `pktbuf` notation to share implementation between plaintext and TLS implementations.

### Motivation

Capture encrypted traffic between a client and a postgres server.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 1]
```diff
@@ -152,4 +152,36 @@ int socket__postgres_process(struct __sk_buff* skb) {
...
+SEC("uprobe/postgres_tls_process")
+int uprobe__postgres_tls_process(struct pt_regs *ctx) {
+    const __u32 zero = 0;
+
+    tls_dispatcher_arguments_t *args = bpf_map_lookup_elem(&tls_dispatcher_arguments, &zero);
+    if (args == NULL) {
+        return 0;
+    }
+
+    // On stack for 4.14
```
Comment (amitslavin) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:164]: what?

File: pkg/network/usm/postgres_monitor_test.go [hunk 2]
```diff
@@ -71,17 +74,53 @@ func (s *postgresProtocolParsingSuite) TestLoadPostgresBinary() {
...
 	serverHost := "127.0.0.1"
 
 	serverAddress := net.JoinHostPort(serverHost, postgresPort)
-	require.NoError(t, postgres.RunServer(t, serverHost, postgresPort, postgres.TLSDisabled))
+	require.NoError(t, postgres.RunServer(t, serverHost, postgresPort, isTLS))
+
+	// With non-TLS, we need to double the stats since we use Docker and the
+	// packets are seen twice. This is not needed in the TLS case since there
+	// the data comes from uprobes on the binary.
+	fixCount := func(count int) int {
```
Comment (amitslavin) [inline pkg/network/usm/postgres_monitor_test.go:109]: maybe we can change from fix to adjust?

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/helpers/pktbuf.h [hunk 3]
```diff
@@ -121,7 +135,7 @@ PKTBUF_READ_BIG_ENDIAN(s8)
             read_into_buffer_##name(buffer, pkt.skb, offset);                                            \
             return;                                                                                      \
         case PKTBUF_TLS:                                                                                 \
-            read_into_user_buffer_##name(buffer, pkt.tls->buffer_ptr + pkt.tls->data_off + offset);      \
+            read_into_user_buffer_##name(buffer, pkt.tls->buffer_ptr + offset);                          \
```
[the remark this answers is not in the document]
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/helpers/pktbuf.h:138]: Just saw you changed it to a different PR so resolving

[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/helpers/pktbuf.h]: To answer your question - it was a bug, we get an offset argument, so we should have used only the offset.
It worked as the code was used only in kafka-tls, and there we never modified data_off field
In postgres TLS I do modify data_off field, and thus the original code resulted in wrong reads from the buffer

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 1 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:164]: elaborated

File: pkg/network/usm/postgres_monitor_test.go (hunk 2 above)
Comment (guyarb) [reply pkg/network/usm/postgres_monitor_test.go]: changed

**Review by amitslavin — APPROVED** [review]

**Pull request merged**